File: alc/funcs.c
// Function handling for the alic compiler
// (c) 2025 Warren Toomey, GPL3

#include "alic.h"
#include "proto.h"

// Given an ASTnode representing a function's name & type
// and a second ASTnode holding a list of parameters, add
// the function to the symbol table. Die if the function
// exists and the parameter list is different or the
// existing function's type doesn't match the new one.
// Return 1 if there was a previous function delaration
// that had a statement block, otherwise 0
int add_function(ASTnode * func, ASTnode * paramlist) {
  Sym *this, *funcptr;
  int paramcnt=0;

  // Try to add the function to the symbol table
  funcptr = add_sym_to(&Symhead, func->strlit, ST_FUNCTION, func->type);

  // The function already exists
  if (funcptr == NULL) {
    // Find the existing prototype
    funcptr = find_symbol(func->strlit);

    // Check the return type
    if (func->type != funcptr->type)
      fatal("%s() declaration has different type than previous: %s vs %s\n",
	    func->strlit, get_typename(func->type),
	    get_typename(funcptr->type));

    // Walk both the paramlist and the member list 
    // in this to verify both lists are the same
    this = funcptr->memb;
    while (1) {
      // No remaining parameters
      if (this == NULL && paramlist == NULL)
	break;

      // Different number of parameters
      if (this == NULL || paramlist == NULL)
	fatal("%s() declaration: # params different than previous\n",
	      func->strlit);

      // Parameter names differ
      if (strcmp(this->name, paramlist->strlit))
	fatal("%s() declaration: param name mismatch %s vs %s\n",
	      func->strlit, this->name, paramlist->strlit);

      // Parameter types differ
      if (this->type != paramlist->type)
	fatal("%s() declaration: param type mismatch %s vs %s\n",
	      func->strlit, get_typename(this->type),
	      get_typename(paramlist->type));

      // Move up to the next parameter in both lists
      this = this->next;
      paramlist = paramlist->mid;
    }

    // All OK. Return the function's intval: 1 means
    // that it was previously declared with a statement block
    return(funcptr->initval.intval);
  }

  // The function is a new one. Walk the parmlist adding
  // each name and type to the function's member list
  for (; paramlist != NULL; paramlist = paramlist->mid) {
    this= add_sym_to(&(funcptr->memb), paramlist->strlit,
		ST_VARIABLE, paramlist->type);
    this->has_addr= false;
    paramcnt++;
  }

  // Set the number of function parameters
  funcptr->count= paramcnt;
  return(0);
}

// Declare a function which has a statement block
void declare_function(ASTnode *f) {
  Sym *this;

  // Add the function declaration to the symbol table.
  // Die if a previous declaration had a statement block
  if (add_function(f, f->left))
    fatal("multiple declarations for %s()\n", f->strlit);

  // Find the function's symbol entry and mark that it
  // does have a statement block
  this= find_symbol(f->strlit);
  this->initval.intval= 1;

  cg_func_preamble(this); new_scope(this);
}

// Generate a function's statement block
void gen_func_statement_block(ASTnode *s) {
  if (Debugfh != NULL) {
    dumpsyms(); dumpAST(s, 0); fflush(Debugfh);
  }

  genAST(s); freeAST(s);
  cg_func_postamble(); end_scope();
}

File: n0c/funcs.py
from typing import List

from cgen import codegen
from defs import Symbol, SymType, ValType, ASTNode, NodeType, fatal
from genast import gen_ast
from syms import add_symbol, find_symbol, set_cur_func


def add_function(node: ASTNode, params: List[Symbol]) -> Symbol:
    name = node.sym.name
    # 检查函数是否已存在
    existing = find_symbol(name)
    if existing:
        if existing.sym_type != SymType.S_FUNC:
            fatal(f"{name} is not a function")
        # 检查参数数量是否匹配
        # existing_params = existing.memb
        # param_count = 0
        # while existing_params:
        #     param_count += 1
        #     existing_params = existing_params.sibling
        # if len(params) != param_count:
        #     fatal(f"Function {name} redefined with different parameter count")
        # # 检查参数类型是否匹配
        # existing_params = existing.memb
        # for i, param in enumerate(params):
        #     if existing_params.type != param.type:
        #         fatal(f"Parameter {i + 1} type mismatch in function {name} redefinition")
        #     existing_params = existing_params.sibling
        return existing

    # 创建新函数符号
    fn_sym = Symbol(name, sym_type=SymType.S_FUNC, val_type=node.val_type)
    fn_sym.params = params
    return fn_sym


def declare_function(node: ASTNode, params: List[Symbol], body: ASTNode) -> ASTNode:
    # 添加函数到符号表
    fn_sym = add_function(node, params)
    # 创建函数节点
    node = ASTNode(NodeType.A_FUNC, left=body)
    node.sym, node.val_type = fn_sym, fn_sym.val_type
    return node


def gen_func_statement_block(fn_sym: Symbol, body: ASTNode) -> None:
    if not fn_sym or not body:
        return
    # 生成函数前导
    codegen.cg_func_preamble(fn_sym)
    # 生成函数体
    gen_ast(body)
    # 生成函数后导
    codegen.cg_func_postamble()
